### PR TITLE
Disable HTTP keep alives

### PR DESCRIPTION
**Pull request by rshriram**
addresses #9113 and other 503 issues.
Rationale: we currently set http keep alive and disable all idle timeouts. So, all connections (between envoys) are perpetual. However, the clients behave differently. Some clients will terminate a connection after set number of requests or will not honor perpetual HTTP connections (i.e. they require a HTTP keep alive should one choose to use it). So these clients will terminate the connection after the preset number of requests or a timer expires. This will cause 503 issues as our Envoys try to send a request on an existing connection.

#12188 tries to address this by adding a retry option for all 503 codes universally. Not many users are comfortable with this option as the default. The alternative, of not setting any retry, will cause  a class of 503 issues as seen by #9113 and related issues.

This PR attempts to disable HTTP keep alives completely. end users wishing to enable it can always use the destination rule to enable the keep alive and control the longevity of the keep alive connection.

Signed-off-by: Shriram Rajagopalan <shriramr@vmware.com>

### CODE REVIEW

**Comment by rshriram** [comment]
addresses https://github.com/istio/istio/issues/12170 and https://github.com/istio/istio/issues/12066

In addition, disable the keep alives so that we dont end up forcing keep alives on all clients. Reduces 503s. Addresses #9113 without the need for retry configs.

**Comment by rshriram** [comment]
cc @knrc / @duderino / @nmittler ..
@robertpanzer / @uk000 to test
supercedes #12216

File: pilot/pkg/networking/core/v1alpha3/cluster.go [hunk 1]
```diff
@@ -993,6 +995,10 @@ func buildDefaultCluster(env *model.Environment, name string, discoveryType apiv
 		Type: discoveryType,
 	}
 
+	// Default to 1 request per connection. This effectively disables HTTP keep-alives for both H1 and H2
+	// and eliminates a bunch of 503 issues
+	cluster.MaxRequestsPerConnection = &types.UInt32Value{Value: 1}
```
Comment (howardjohn) [inline pilot/pkg/networking/core/v1alpha3/cluster.go]: Why is this in the same change? Seems unrelated?

File: pilot/pkg/model/push_context.go [hunk 2]
```diff
@@ -502,11 +502,21 @@ func (ps *PushContext) DestinationRule(proxy *Proxy, service *Service) *Config {
...
-			return ps.namespaceLocalDestRules[proxy.ConfigNamespace].destRule[host].config
+	// TODO: once we move default DestinationRule to root config namespace,
+	// remove the hacky check below.  Until then, we need an exception for
+	// istio-system namespace, as it happens to be the home of global dest
+	// rules for mTLS, as well as the home of the gateway proxy. The root
+	// config namespace is technically not supposed to have any proxy
+	// istio-ingressgateway and istio-egressgateway wont hit this if block.
+	// instead, dest rules will first be looked up from the service's own namespace
+	// if not found, we look at ALL dest rules across all namespaces. This is the old behavior
+	if proxy.ConfigNamespace != IstioSystemNamespace {
```
Comment (costinm) [inline pilot/pkg/model/push_context.go]: ConfigNamespace should be istio-system, never istio-system. When would this happen ?

File: pilot/pkg/networking/core/v1alpha3/cluster.go [hunk 3]
```diff
@@ -980,6 +980,8 @@ func buildDefaultPassthroughCluster() *apiv2.Cluster {
 		Type:           apiv2.Cluster_ORIGINAL_DST,
 		ConnectTimeout: 1 * time.Second,
 		LbPolicy:       apiv2.Cluster_ORIGINAL_DST_LB,
+		// Disable keep alive
+		MaxRequestsPerConnection: &types.UInt32Value{Value: 1},
```
Comment (costinm) [inline pilot/pkg/networking/core/v1alpha3/cluster.go]: This has huge impact on performance. At least protect it with a flag/feature, we can't hardcode it to 1.

@mandarjog

File: pilot/pkg/networking/core/v1alpha3/cluster.go (hunk 1 above)
Comment (costinm) [inline pilot/pkg/networking/core/v1alpha3/cluster.go]: Same - did you do any benchmark with the impact ? Numbers ?

File: pilot/pkg/model/push_context.go [hunk 4]
```diff
@@ -521,8 +531,20 @@ func (ps *PushContext) DestinationRule(proxy *Proxy, service *Service) *Config {
...
+	// if and only if there is no config root namespace defined. If a config root namespace is defined,
+	// only look for the destination rule from the config root namespace
+	if ps.Env != nil && ps.Env.Mesh != nil && ps.Env.Mesh.RootNamespace != "" {
+		if ps.namespaceExportedDestRules[ps.Env.Mesh.RootNamespace] != nil {
+			if host, ok := MostSpecificHostMatch(service.Hostname,
+				ps.namespaceExportedDestRules[ps.Env.Mesh.RootNamespace].hosts); ok {
+				return ps.namespaceExportedDestRules[ps.Env.Mesh.RootNamespace].destRule[host].config
+			}
+		}
+	} else {
```
Comment (costinm) [inline pilot/pkg/model/push_context.go]: else if (! hasSidecar) - if it has sidecar, don't look anywhere else.

**Comment by costinm** [comment]
I don't mind an option (off by default) to do this - but given the huge impact on both performance and infrastructure it must be off by default.  Doing TLS handshake on each request is quite horrible, and even with TCP it's bad.

**Comment by costinm** [comment]
Also for H2 is even worse - it effectively disables H2... The main point of H2 is to multiplex and better reuse connections.

File: pilot/pkg/networking/core/v1alpha3/cluster.go (hunk 1 above)
Comment (duderino) [reply pilot/pkg/networking/core/v1alpha3/cluster.go]: I don't think we can ignore this.  We'd need performance and efficiency numbers before making this sort of change
